### PR TITLE
Ingester: fix  wrong conversion of global into local limits

### LINKED ISSUE #4208: Scaling down ingesters in 1 zone at a time conflicts with "max series per user" limit
The "max series per user" limit, enforced in ingesters, is implemented as a per-ingester local limit. The local limit is configured equal to the global limit divided by the number of running ingesters in the tenant's shard (if shuffle sharding is configured for the tenant):
https://github.com/grafana/mimir/blob/5d158054ca1bdddc953e42ef80ef61589f8b4795/pkg/ingester/limiter.go#L208-L234

The limit is not really zone-aware, in the sense it doesn't consider the case the number of ingesters per-zone could be unbalanced. In a normal condition, we expect the number of ingesters per zone to be balanced.

However, in the event of scaling down ingesters, 1 zone at a time, the number of per-zone ingesters will be different until the scale down is completed. This means that in the scaled down zone, the computed local limit may be incorrect (too low) because it assumes an equally distributed number of ingesters per zone, while it's not.

_This issue originates from a real issue I've experienced while scaling down a production Mimir cluster._

### PR DESCRIPTION
**Pull request by duricanikolic**
#### What this PR does
This PR fixes a bug introduced by issue #4208: conversion of global to local limits didn't take into account the actual number of ingesters in the zone of interest, but assumed that ingesters were equally distributed in each zone. As a direct consequence, scaling down ingesters in a certain zone caused the limits of that zone to be too low. This PR fixes that problem.
For example, let's assume the following:
- global limit (`GL`) is 1M, 
- replication factor (`RF`) is 3, 
- zone count (`ZC`) is 3, for example `zone-a`, `zone-b`, `zone-c`
- number of ingesters per zone (`IA` = `IB` = `IC`) is 10

With the previous approach local limit would be calculated, for all 3 zones, as 
```
GL * RF / (IA + IB + IC) = 1M * 3 / 30 = 100K.
```
Similarly, the new approach would calculate the same result for each zone 
```
GL * RF / ZC / IA = GL * RF / ZC / IB = GL * RF / ZC / IC = 100K,
```
Suppose that we now scale down ingesters in `zone-a` (`IA`) to 5. The previous approach would, again, assign the same local limit to all 3 zones to
```
GL * RF / (IA + IB + IC) = 1M * 3 / (5 + 10 + 10) = 120K,
```
which is wrong. The right limits are calculated by the new approach like this:
- for `zone-a`: ```GL * RF / ZC / IA = 1M * 3 / 3 / 5 = 200K```
- for `zone-b` and `zone-c`: ```GL * RF / ZC / IB = GL * RF / ZC / IC = 1M * 3 / 3 / 10 = 100K``` .

#### Which issue(s) this PR fixes or relates to

Fixes #4208 

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:230]: I think we need to keep using `ShuffleShardExpectedInstances`. This protects from the case the shard size has been configured to a value not divisible by the number of zones. In this case we round up.

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:213]: I wouldn't read it anymore, but use the value of `getHealthyIngestersInZoneCount()`. If it's 0, then we return `0` from here. Also to avoid race conditions, we should read `getHealthyIngestersInZoneCount()` only once and then use that value in the rest of the function.

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:243]: Based on the comment above, this code should be executed also if zone-aware replication is enabled. If enabled, you should call `ShuffleShardExpectedInstancesPerZone()` otherwise you should call `ShuffleShardExpectedInstances()`.

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:224]: To reduce likelihood of race conditions, I would call `getZonesCount()` only once and then use in the rest of the logic.

**Review by pracucci — COMMENTED** [review]
Thanks for working on this! I left some initial comments, cause I think the logic is not correct in case of zone-aware replication enabled **and** shuffle sharding (shard size is set).

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:31]: Please list which limits are affected.

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:220]: I think you can remove this whole code block. You already have `if ingestersInZoneCount == 0 {` below.

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:254]: [nit] I think there's a typo here: "zoneInside" and "," at the end of the sentence.

**Review by pracucci — APPROVED** [review]
The logic LGTM, thanks! I left few last nits 🙏 Since this logic is tricky and it's in the critical path, I would like another pair of 👀 : @pstibrany could you review it too, please?

**Comment by duricanikolic** [comment]
> The logic LGTM, thanks! I left few last nits 🙏 Since this logic is tricky and it's in the critical path, I would like another pair of 👀 : @pstibrany could you review it too, please?

The logic is indeed tricky, and I tried to cover all possible cases with a separate test. If @pstibrany finds a case that is not covered by the logic, we need to add the corresponding test as well

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:214]:
```suggestion
```

File: pkg/ingester/limiter.go
Comment (pracucci) [inline pkg/ingester/limiter.go:231]:
```suggestion
		ingestersInZoneCount = l.ring.HealthyInstancesCount()
```

File: pkg/ingester/limiter_test.go
Comment (pstibrany) [inline pkg/ingester/limiter_test.go:150]: In that case, total number of ingesters (`ringIngesterCount`) should be 8, not 9.

File: pkg/ingester/limiter_test.go
Comment (pstibrany) [inline pkg/ingester/limiter_test.go:170]: `ringIngesterCount` doesn't reflect that.

File: pkg/ingester/limiter_test.go
Comment (pstibrany) [inline pkg/ingester/limiter_test.go:194]: I'd suggest adding test cases for scaling-up too. That's also a scenario where limits are currently wrong.

**Review by pstibrany — COMMENTED** [review]
Looks good, but I have a question about "healthiness", and suggestion for scale-up test cases.

File: pkg/ingester/limiter_test.go
Comment (duricanikolic) [reply pkg/ingester/limiter_test.go:194]: @pracucci has also mentioned the problem of scaling-up. I am not sure if this PR also fixes the scaling-up problem. Logically, I would say yes: we now know how many healthy ingesters are there in each zone, and we don't estimate that number like we used to do in the previous implementation. But I might miss some real-life use case. What do you think @pracucci @pstibrany?

Comment (pstibrany) [reply pkg/ingester/limiter_test.go:194]: Yes, I think the problem is the same... when scaling up, say from 10+10+10 (30 ingesters in total) configuration to 15+10+10 (35 ingesters in total), previously this affected limits of ingesters in all zones. Now it will affect only scaled-up zone. I'd recommend adding a test showing that (one test for zone that's affected by scale up, one test for zone that's not affected by scale up).

File: pkg/ingester/limiter_test.go
Comment (duricanikolic) [reply pkg/ingester/limiter_test.go:152]: I am not sure I can answer this question. We would need @pracucci 's support. When I added the [HealthyInstancesInZoneCount()](https://github.com/grafana/dskit/blob/main/ring/lifecycler.go#L389) function to Lifecycler, I initially counted all instances, but then @pracucci suggested that we should count only the healthy ones, so I implemented it [this way](https://github.com/grafana/dskit/blob/main/ring/lifecycler.go#L819-L822).

File: pkg/ingester/limiter_test.go
Comment (duricanikolic) [reply pkg/ingester/limiter_test.go:194]: @pstibrany Shall I also add tests showing that limits do not change in the zones not affected by scaling down? Just for completeness

Comment (pstibrany) [reply pkg/ingester/limiter_test.go:194]: Sounds good to me

File: pkg/ingester/limiter_test.go
Comment (pstibrany) [reply pkg/ingester/limiter_test.go:152]: Regarding limits and healthy ingesters… we have a discussion whether using healthy ingesters only makes sense. Let’s consider an example. User has 1M series limit. We have 3 zones, each with 10 ingesters, and we use RF=3. So each ingester has 100k limit.

During rollout, we stop 5 ingesters at once in single zone. So there will be only 5 healthy ingesters left in the zone.
At this point limits at each remaining ingester in the same zone is recomputed to allow 200k series.
Mimir now always uses “WriteNoExtend”, so if user wasn’t hitting the 100k limit yet, this is fine. But if user was hitting 100k limit before, now we temporarily allow twice as many series for the user. After rollout in the zone finishes, per-ingester limit will be set back to 100k, but series will be in memory already.

Situation is similar when shuffle sharding is in use for the user. Let’s say user has 24 ingesters in total (ie. 8 per zone).
Limit per ingester if all ingesters are healhty will be 125k. During rollout we restart 5 ingesters in one zone, and only 5 will be healthy. Since 5 is less than 8 requested ingesters from the zone, temporary limit on each ingester will be 200k again.

Note that shuffle shard doesn’t take healthiness into consideration, so we’re mixing unhealthy with healthy ingesters when doing the check.
Given that we always use WriteNoExtend, I am not sure why we only consider healthy ingesters when computing the limit.

Comment (pracucci) [reply pkg/ingester/limiter_test.go:152]: Talked to Peter on Slack. I think the reason why we're not seeing it happening is because the ring client skips the entire zone if there are unhealthy instances, so we don’t write at all to ingesters in the zone we’re rollout out.

However, I'm 👍 with the change proposed by Peter, to count all instances not just healthy ones, given it looks more correct to me too.

Comment (duricanikolic) [reply pkg/ingester/limiter_test.go:152]: Does it mean we need to do the following things?

1. update `dskit` by replacing its current `HealthyInstancesInZoneCount()` by `InstancesInZoneCount()` and by updating its implementation
2. once it is merged, update this PR and use the new `InstancesInZoneCount()` method instead? Beside some syntactical changes (replacing "healthyIngesters" with "ingesters"), I don't see other things that should be changed here.

Comment (pstibrany) [reply pkg/ingester/limiter_test.go:152]: That looks correct to me.

Comment (pracucci) [reply pkg/ingester/limiter_test.go:152]: I think you also need to add `InstancesCount()` in dskit to use when the zone-aware replication is disabled.

Comment (duricanikolic) [reply pkg/ingester/limiter_test.go:152]: @pstibrany @pracucci Could you please review https://github.com/grafana/dskit/pull/270?

File: pkg/ingester/limiter.go
Comment (pstibrany) [inline pkg/ingester/limiter.go:216]: We don't need this function... `ShuffleShardExpectedInstancesPerZone` works fine for both cases.

**Review by pstibrany — APPROVED** [review]
Thanks!

**Comment by pstibrany** [comment]
This PR needs to be rebased before merging.

**Comment by pracucci** [comment]
@duricanikolic Look at the PR's diff. As you can see there are changes not related to this PR. I think you merged instead of rebase. You should rebase `main`. Since you have a bunch of commits, the easiest way to rebase reducing conflicts may be squash all your commits into 1 commit, and then `git rebase main`.

**Comment by duricanikolic** [comment]
> 

Hi @pracucci, 
I have updated my PR

File: pkg/ingester/limiter.go
Comment (pracucci) [reply pkg/ingester/limiter.go:217]: Sorry for writing in italian. Let me translate it:

 I don’t think it’s correct to do the division operations in integer, because with each operation you lose precision. This is why we previously performed these operations in floating point and then converted the result to int.

Comment (duricanikolic) [reply pkg/ingester/limiter.go:217]: This was requested by @pstibrany. I will revert it so we ensure that we don't lose the precision

Comment (pstibrany) [reply pkg/ingester/limiter.go:217]: translation:

> I don't think it's correct to do the division operations in integer, because with each operation you lose precision. This is why we first performed these operations in floating point and then converted the result to int.

Comment (pstibrany) [reply pkg/ingester/limiter.go:217]: I suggested to either simplify that expression and remove floats, or add a test to show why floats are required. While I agree about loss of precision, I doubt it will have any meaningful effect here in practice.

Comment (pracucci) [reply pkg/ingester/limiter.go:217]: It's probably not meaningful because our limits are very high numbers, but I don't understand we shouldn't be correct.

Comment (pstibrany) [reply pkg/ingester/limiter.go:217]: Max difference between using floats or not is `1 / ingestersInZoneCount`.

Comment (pracucci) [reply pkg/ingester/limiter.go:217]: Ok, then I just suggest to add a comment explaining why we do the int division, so that the next code reader won't ask themselves the same question.

Comment (pracucci) [reply pkg/ingester/limiter.go]: @duricanikolic I think Peter feels stronger than me on this, so I would get stick to what Peter suggested previously (int division). I would just add a comment explaining why it's not an issue ("Max difference between using floats or not is 1 / ingestersInZoneCount").

**Review by pracucci — APPROVED** [review]
LGTM! (modulo the nit on precision, sorry for the back and forth on this)

File: pkg/ingester/limiter.go
Comment (duricanikolic) [reply pkg/ingester/limiter.go]: A question for both @pracucci and @pstibrany: If we have a global limit 1000, replication factor 2 and number of zones 3, and let's say 1 ingester in that zone, should we return 666 or 667 as the result (1000 * 2 / 3)?

Comment (pracucci) [reply pkg/ingester/limiter.go]: Answer is: "doeasn't matter". There's no practical difference between the two. Which is Peter's point. I think int divisions are always floored (like casting float to int) so probably it's 666.

Comment (pstibrany) [reply pkg/ingester/limiter.go]: > @duricanikolic I think Peter feels stronger than me on this, so I would get stick to what Peter suggested previously (int division).

No need to revert back.


> A question for both @pracucci and @pstibrany: If we have a global limit 1000, replication factor 2 and number of zones 3, and let's say 1 ingester in that zone, should we return 666 or 667 as the result (1000 * 2 / 3)?

Both are fine. Returning ceil would benefit user -- allow them to write one extra series per ingester.

Comment (duricanikolic) [reply pkg/ingester/limiter.go]: OK, so today I am going to test this PR on mimir-dev-09, and once it is done I will let you know, so we can merge this.

**Review by osg-grafana — APPROVED** [review]

**Pull request merged**
